### PR TITLE
feat: Adds context trace information to error log entries if there's any.

### PR DESCRIPTION
**Pull request by amanda-tarafa**
As always, one commit at a time.

### CODE REVIEW

**Comment by amanda-tarafa** [comment]
Diff errors are the ones we accepted from before.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/LogEntryExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            }
+
+            // If there's currently a Google trace and span use that one.
+            // This means that the Google Trace component of the diagnostics library
+            // has been initialized.
+            // Else attempt to use an external trace context.
+            if ((ContextTracerManager.GetCurrentTraceContext() ?? serviceProvider?.GetService<ITraceContext>()) is ITraceContext context
+                && context.TraceId is string)
+            {
+
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/LogEntryExtensions.cs:42]: Nit: remove blank line

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/LogEntryExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            if (traceTarget is null)
+            {
+                return;
+            }
+
+            // If there's currently a Google trace and span use that one.
+            // This means that the Google Trace component of the diagnostics library
+            // has been initialized.
+            // Else attempt to use an external trace context.
+            if ((ContextTracerManager.GetCurrentTraceContext() ?? serviceProvider?.GetService<ITraceContext>()) is ITraceContext context
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/LogEntryExtensions.cs:39]: I think I'd probably extract this as a local variable first:

```csharp
var context = ContextTracerManager.GetCurrentTraceContext() ?? serviceProvider?.GetService<ITraceContext>();
if (context is ITraceContext traceContext)
{
   ...
}
```

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/ErrorReporting/ErrorReportingContextExceptionLogger.cs [hunk 3]
```diff
@@ -158,14 +170,18 @@ private LogEntry CreateReportRequest(Exception exception, IContextWrapper contex
...
+            var entry = new LogEntry
             {
                 Resource = _options.EventTarget.MonitoredResource,
                 LogName = _logName,
                 Severity = LogSeverity.Error,
                 Timestamp = timestamp,
                 JsonPayload = errorEvent
             };
+
+            entry.SetTraceAndSpanIfAny(_traceTarget, _serviceProvider, null);
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/ErrorReporting/ErrorReportingContextExceptionLogger.cs:182]: If you made the extension method return the entry, you could have this as:

```csharp
return new LogEntry
{
    Resource = _options.EventTarget.MonitoredResource,
    LogName = _logName,
    Severity = LogSeverity.Error,
    Timestamp = timestamp,
    JsonPayload = errorEvent
}.SetTraceAndSpanIfAny(_traceTarget, _serviceProvider, null);
```

Just a thought - not particularly bothered either way.

**Review by jskeet — APPROVED** [review]
As ever with diagnostics, I may be missing something - but I *think* it's fine. Just a few nits, and they're all optional...

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/LogEntryExtensions.cs (hunk 2 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/LogEntryExtensions.cs:39]: Done

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/LogEntryExtensions.cs (hunk 1 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/LogEntryExtensions.cs:42]: Done

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/ErrorReporting/ErrorReportingContextExceptionLogger.cs (hunk 3 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/ErrorReporting/ErrorReportingContextExceptionLogger.cs:182]: Ah yes, will do that.

**Review by amanda-tarafa — COMMENTED** [review]
Pushing shortly with all your comments addressed. I rebased in a single commit now, and I'm re-running all tests on local. I'll merge when it's green. Thanks!

**Pull request merged**